### ISSUE TITLE
Add tests to all routes

### ISSUE THREAD

**Issue by mmmaaatttttt** [rithmschool/numbers_api#5]
Currently no routes have tests. Let's make sure the app is well-tested.

Estimated points: 6

**Comment by mmmaaatttttt** [comment]
important thing to test: make sure `json` key in the query string does what it's supposed to. had an issue where the major version bump from express 2 to express 4 broke this functionality (because of a now deprecated reference to the `req.param()` function)

**Comment by stephanedreyfus** [comment]
👍 (I need the testing practice, so will probably be asking for some guidance on these.)

**Comment by mmmaaatttttt** [comment]
@stephanedreyfus @stephenisau @WinnieGotYu @kathyn262 Can the four of you split into two groups of two and start tackling this in pairs? It's probably worth breaking this into a couple of issues -- I'd start by listing all the JS files that need tests and then figuring out who will work on testing which files.

**Comment by stephanedreyfus** [comment]
Hi everyone! Happy to be working on testing with you. I'm from Rithm-10 and
have been working as a TA at Springboard for a little while now. I'm not
the best at testing yet, so this is going to be a bit slow for me at the
beginning. Nonetheless, looking forward to partnering with someone and
getting some testing done!

Stéphane

**Comment by mmmaaatttttt** [comment]
In response to some questions I've received on the tests:

- in `app.js`, for now let's add integration tests for the routes on L126 and 151. eventually when that commented out code comes back we'll need more tests, but i think this is fine for now.
- re: `models/facts`, every function in here should be unit tested. we should also do an audit of what's provided by underscore -- it's possible we're writing methods that come out of the box in that library. 
- re: `models/data`, let's make sure every function in here is unit tested as well. some other things to consider: (1) there are two giant functions in here, if we can break them up they'll be easier to reason about and easier to test. (2) `_.each` is cool if you have an object, but if we're working with arrays let's just use `.forEach`

**Comment by mmmaaatttttt** [comment]
@stephenisau @stephanedreyfus @kathyn262 @WinnieGotYu -- what's the status of this issue?

**Comment by stephanedreyfus** [comment]
I let the group know that I was being too ambitious when I signed up to
help. For multiple reasons I will be just observing unless I suddenly find
myself with a great deal more free time. I apologized to the group but I
should apologize to you as well Matt: sorry! In short, I don't know what
other progress has been made on this issue.

**Comment by kathyn262** [comment]
@mmmaaatttttt Overall, we’re still working on it! It took some time to understand the code, but we’re writing tests now. @WinnieGotYu and I are working on `models/facts`. @stephenisau is working on `models/data`.

**Comment by stephenisau** [comment]
Hey @mmmaaatttttt! Slowly making progress on the tests, I have been a bit busy with personal life so haven't had much of a chance to consistently work on the tests. @stephanedreyfus mentioned he'd be taking a bit of a backseat on this ticket so i'll do my best to update him once I make some progress.

The 2 big functions in the file, `reader_norm`  and `reader_manual` seem to be reading from .txt files and doing some data cleaning using the third function `normalize_common`. Two questions came up that i'm not quite sure how to answer, hoping to get some clarity on the two!

1. The two functions read inside directories using`models/` as the base directory. Would it make sense to create a test directory with their own data files inside so the functions within the tests have a "single source of truth"? (ie. I create a `tests/data` directory and point to those files in our tests to read data from.)

2. The two main functions rely heavily on logging errors to the console if they aren't able to read the data from the .txt files. I was planning on writing a few tests that expect these failures to happen, but don't know how to test outputs from console.logs inside a function. I came across this stackoverflow post regarding testing logs to the console (https://stackoverflow.com/questions/49096093/how-do-i-test-a-jest-console-log) but not sure if that's the right way to go about it. I might be thinking about this wrong but hoping to get a better understanding on how to write tests for this part.

Any clarity on these two points would be much appreciated, thank you!!

**Comment by mmmaaatttttt** [comment]
Hey @stephenisau -- sorry for the delay. Hopefully some of the discussion from Monday's OH is relevant here. some thoughts:

1. Depending on how much information you're trying to mock, you could either do something like:

```js
let fsMock = jest.spyOn(fs, "readDirSync")
fsMock.mockImplementation( /* Your implementation here */);
```

and similarly for readFileSync. But if you're trying to mock a larger text file, I'd probably follow the pattern of having a `__mocks__` directory. The jest docs may be helpful here, since i think their example mocks exactly what you're trying to mock: the fs module: https://jestjs.io/docs/en/manual-mocks

2. Re: mocking console.log, the stackoverflow you found is good, I'd probably follow the pattern described in [this answer](https://stackoverflow.com/a/59225389/4930688). Don't forget to restore the mock after the test, or in an afterEach!

**Comment by mmmaaatttttt** [comment]
closed by #39
